server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java

```diff
@@ -0,0 +1,146 @@
+            @Override
+            public void onResponse(Void aVoid) {
+                try {
+                    // TODO: Propagate the connection to the listener
+                    Transport.Connection newConnection = connectionManager.getConnection(node);
+                    addConnection(newConnection);
+                    newConnection.addCloseListener(ActionListener.wrap(() -> removeConnection(newConnection)));
+                } catch (NodeNotConnectedException e) {
+                    listener.onFailure(e);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    public boolean nodeConnected(DiscoveryNode node) {
+        return connectionManager.nodeConnected(node);
+    }
+
+    public void openConnection(DiscoveryNode node, ConnectionProfile profile, ActionListener<Transport.Connection> listener) {
+        connectionManager.openConnection(node, profile, listener);
+    }
+
+    public Transport.Connection getRemoteConnection(DiscoveryNode node) {
+        try {
+            return connectionManager.getConnection(node);
+        } catch (NodeNotConnectedException e) {
+            return new ProxyConnection(getAnyRemoteConnection(), node);
+        }
+    }
+
+    public Transport.Connection getAnyRemoteConnection() {
+        List<Transport.Connection> localConnections = this.connections;
+        if (localConnections.isEmpty()) {
+            throw new NoSuchRemoteClusterException(clusterAlias);
+        } else {
+            long curr;
+            while ((curr = counter.incrementAndGet()) == Long.MIN_VALUE);
+            return localConnections.get(Math.floorMod(curr, localConnections.size()));
+        }
+    }
+
+    public ConnectionManager getConnectionManager() {
+        return connectionManager;
+    }
+
+    public int size() {
+        return connectionManager.size();
+    }
+
+    public void close() {
+        connectionManager.closeNoBlock();
+    }
+
+    private synchronized void addConnection(Transport.Connection addedConnection) {
+        ArrayList<Transport.Connection> newConnections = new ArrayList<>(this.connections);
+        newConnections.add(addedConnection);
+        this.connections = Collections.unmodifiableList(newConnections);
+    }
+
+    private synchronized void removeConnection(Transport.Connection removedConnection) {
+        int newSize = this.connections.size() - 1;
+        ArrayList<Transport.Connection> newConnections = new ArrayList<>(newSize);
+        for (Transport.Connection connection : this.connections) {
+            if (connection.equals(removedConnection) == false) {
+                newConnections.add(connection);
+            }
+        }
+        assert newConnections.size() == newSize : "Expected connection count: " + newSize + ", Found: " + newConnections.size();
```

Comment:
We can sometimes get two close events if we have two concurrent calls to `connectToNode` for the same node which give us the same connection object. Only one of these close calls removes the connections, the other becomes a NOOP and trips this assertion.